### PR TITLE
Plugins: Angular deprecation: Fix AngularDeprecationNotice not being rendered on first page load

### PR DESCRIPTION
**Pull request by xnyo**
**What is this feature?**

Fixes an issue where the Angular warning at the top of the dashboards would not appear until entering/exiting editing mode if there are Angular panel plugins in the dashboard (but not Angular data sources).

This happens because the `plugins` property (which is used to test if Angular is detected on the panel) is populated asynchronously, so angular is not detected on the first render:
https://github.com/grafana/grafana/blob/74c0463cd347e8b7a239dbb36ec6a2ef4926bc8e/public/app/features/dashboard/state/PanelModel.ts#L630-L634

This PR replaces that check by testing for the `angularDetected` and `hideAngularDeprecation` in `config` (`grafanaBootData`), which is always available unlike the `plugin` property for a panel.

The feature flag is enabled by default in v10.4.0, so this PR will be backported to Grafana v10.4.0 as well.

**Why do we need this feature?**

Bugfix.

**Who is this feature for?**

Grafana users with Angular panel plugins.

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/dashboard/state/DashboardModel.ts [hunk 1]
```diff
@@ -1339,16 +1340,29 @@ export class DashboardModel implements TimeModel {
...
 
-  hasAngularPlugins(): boolean {
-    return this.panels.some((panel) => {
+  async hasAngularPlugins(): Promise<boolean> {
+    for (const panel of this.panels) {
+      let plugin;
+      if (!panel.plugin) {
+        try {
+          plugin = await dispatch(loadPanelPlugin(panel.type));
+          await panel.pluginLoaded(plugin);
```
Comment (academo) [inline public/app/features/dashboard/state/DashboardModel.ts:1349]: this logic as is written is kind of blocking, waiting in this line for each of the panels in the loop, is this your intention?

an alternative way of doing this without blocking waiting for each plugin to individually load but instead let the js engine handle the concurrency (if available):

```typescript
async hasAngularPlugins(): Promise<boolean> {
  await Promise.all( this.panels.map ( (panel) => {
     if (panel.plugin) { return } // Promise.all handle non-promises fine
     try {
        const  plugin = await dispatch(loadPanelPlugin(panel.type));
        //this is the promise we await for
        return panel.pluginLoaded(plugin);
     } catch(e) { return }
  });
  // ...
```

File: public/app/features/dashboard/state/DashboardModel.ts [hunk 2]
```diff
@@ -1339,16 +1340,29 @@ export class DashboardModel implements TimeModel {
     return this.originalDashboard;
   }
 
-  hasAngularPlugins(): boolean {
-    return this.panels.some((panel) => {
+  async hasAngularPlugins(): Promise<boolean> {
+    for (const panel of this.panels) {
+      let plugin;
```
Comment (academo) [inline public/app/features/dashboard/state/DashboardModel.ts:1345]: nit: this `plugin` variable should be a constant in line 1348

File: public/app/features/dashboard/state/DashboardModel.ts (hunk 1 above)
Comment (xnyo) [reply public/app/features/dashboard/state/DashboardModel.ts:1352]: Good catch! It's not necessary to wait for each plugin in the loop, I changed it to `Promise.all` :+1:

File: public/app/features/dashboard/state/DashboardModel.ts (hunk 2 above)
Comment (xnyo) [reply public/app/features/dashboard/state/DashboardModel.ts]: Fixed!

File: public/app/features/dashboard/state/DashboardModel.ts [hunk 3]
```diff
@@ -1339,7 +1340,23 @@ export class DashboardModel implements TimeModel {
...
+  async hasAngularPlugins(): Promise<boolean> {
+    // Ensure all panel plugins are loaded before checking if they are Angular plugins
+    await Promise.all(
+      this.panels.map(async (panel) => {
+        if (panel.plugin) {
+          return;
+        }
+        try {
+          const plugin = await dispatch(loadPanelPlugin(panel.type));
+          await panel.pluginLoaded(plugin);
```
Comment (academo) [inline public/app/features/dashboard/state/DashboardModel.ts:1352]: this should be return, not await.

Comment (xnyo) [reply public/app/features/dashboard/state/DashboardModel.ts:1352]: Ah, makes sense. However, what about the `await dispatch` above it? Is it fine if we `await` it and `return` only the second promise?

**Review by mckn — COMMENTED** [review]
This is good but it can be done slightly different and become great. If you add the check wether or not dashboard contains angular panels in redux you won't do the async things you are doing in the page since the page will be updated when the redux state changes.

I have not worked in this code for some time but I think the following would work:

- Add a new case in the dashboard slice and a new property indicating if the dashboard as angular panels. E.g. `hasAngularPanels = false;`.
- Connect this property in the DashboardPage so the page will be updated if that part of the state changes.
- When each panel is initialised we dispatch the `initPanelState` action (I believe). That action It is an async thunk so the plugin is fetched and you have access to the meta data.
- Add a statement in that thunk action where you check if the panel is an angular panel. If it is an angular panel dispatch the new action created in the dashboard slice.

That is it. If you do it like this it should work. I'm happy to pair on this if needed.

File: public/app/features/dashboard/state/DashboardModel.ts [hunk 4]
```diff
@@ -1339,7 +1340,23 @@ export class DashboardModel implements TimeModel {
...
-  hasAngularPlugins(): boolean {
+  async hasAngularPlugins(): Promise<boolean> {
+    // Ensure all panel plugins are loaded before checking if they are Angular plugins
+    await Promise.all(
+      this.panels.map(async (panel) => {
+        if (panel.plugin) {
+          return;
+        }
+        try {
+          const plugin = await dispatch(loadPanelPlugin(panel.type));
```
Comment (jackw) [inline public/app/features/dashboard/state/DashboardModel.ts:1351]: Now sure how big a deal this is (as from what I can see this isn't render blocking) but would like to point out the current loading mechanism only loads panel plugins when the panel comes into view (i.e. as a user scrolls). This change means all panel plugins for a dashboard are loaded up front now.

File: public/app/features/dashboard/state/DashboardModel.ts [hunk 5]
```diff
@@ -1339,7 +1340,23 @@ export class DashboardModel implements TimeModel {
     return this.originalDashboard;
   }
 
-  hasAngularPlugins(): boolean {
+  async hasAngularPlugins(): Promise<boolean> {
+    // Ensure all panel plugins are loaded before checking if they are Angular plugins
```
Comment (jackw) [inline public/app/features/dashboard/state/DashboardModel.ts:1344]: Do we need to load the plugins to know if they are angular plugins? I thought the information was already available in `config.panels`. Couldn't we use that?

File: public/app/features/dashboard/state/DashboardModel.ts (hunk 3 above)
Comment (academo) [reply public/app/features/dashboard/state/DashboardModel.ts]: ah, missed that await, then it'll be blocking anyway, in this case you should

```
return const dispatch(loadPanelPlugin(panel.type)).then( plugin => panel.pluginLoaded(plugin))
```

to return a single promise.

File: public/app/features/dashboard/state/DashboardModel.ts (hunk 5 above)
Comment (xnyo) [reply public/app/features/dashboard/state/DashboardModel.ts]: You're right, I wanted to use `panel.isAngularPlugin()` across all warnings for consistency, but by accessing `grafanaBootData` makes the PR much simpler, plus it solves the issue about having to load all plugins on the first page load you were talking about in your other comment.

I have pushed some changes so it uses `grafanaBootData` instead

**Comment by xnyo** [comment]
I have simplified the PR following Jack's suggestion about using `config` rather than loading all plugins on the first page load. Please take a look :smile:

**Review by mckn — APPROVED** [review]
LGTM! Awesome work on this 🙌🏻 Less is more and this is for sure the best solution.

**Review by jackw — APPROVED** [review]
LGTM! 🚀

**Pull request merged**